src/XMonad/Operations.hs

```diff
@@ -442,20 +443,78 @@ initColor dpy c = C.handle (\(C.SomeException _) -> return Nothing) $
 
 ------------------------------------------------------------------------
 
+-- | A type to help serialize xmonad's state to a file.
+data StateFile = StateFile
+  { sfWins :: W.StackSet  WorkspaceId String Window ScreenId ScreenDetail
+  , sfExt  :: [(String, String)]
+  } deriving (Show, Read)
+
+-- | Write the current window state (and extensible state) to a file
+-- so that xmonad can resume with that state intact.
+writeStateToFile :: X ()
+writeStateToFile = do
+    let maybeShow (t, Right (PersistentExtension ext)) = Just (t, show ext)
+        maybeShow (t, Left str) = Just (t, str)
+        maybeShow _ = Nothing
+
+        wsData   = W.mapLayout show . windowset
+        extState = catMaybes . map maybeShow . M.toList . extensibleState
+
+    path  <- stateFileName
+    stateData <- gets (\s -> StateFile (wsData s) (extState s))
+    catchIO (writeFile path $ show stateData)
+
+-- | Read the state of a previous xmonad instance from a file and
+-- return that state.
+readStateFile :: (LayoutClass l Window, Read (l Window)) => XConfig l -> X (Maybe XState)
+readStateFile xmc = do
+    path <- stateFileName
+    raw  <- userCode $ io (readFile path <* removeFile path)
```

Comment:
In any case I don't understand why it needs to be removed at all.  Seems better to just let it hang around until it is overwritten on the next restart.  That way if xmonad crashes the state can still be restored from the most recent restart.